integration-cli/docker_cli_daemon_test.go

```diff
@@ -152,6 +153,79 @@ func (s *DockerDaemonSuite) TestDaemonStartIptablesFalse(c *check.C) {
 	}
 }
 
+// Make sure we cannot shrink base device at daemon restart.
+func (s *DockerDaemonSuite) TestDaemonRestartWithInvalidBasesize(c *check.C) {
+	s.d.storageDriver = "devicemapper"
+	if err := s.d.Restart(); err != nil {
+		c.Fatalf("we should have been able to start the daemon: %v", err)
+	}
+
+	oldBasesizeBytes := s.getBaseDeviceSize(c)
+	newBasesizeBytes := 1073741824 //1GB in bytes
+
+	if int64(newBasesizeBytes) < oldBasesizeBytes {
+		if err := s.d.Restart("--storage-opt", fmt.Sprintf("dm.basesize=%d", newBasesizeBytes)); err == nil {
+			c.Fatalf("daemon should not have started as new base device size is less than existing base device size: %v", err)
+		}
+	}
+}
+
+// Make sure we can grow base device at daemon restart.
+func (s *DockerDaemonSuite) TestDaemonRestartWithIncreasedBasesize(c *check.C) {
+	s.d.storageDriver = "devicemapper"
+	if err := s.d.Restart(); err != nil {
+		c.Fatalf("we should have been able to start the daemon: %v", err)
+	}
+
+	oldBasesizeBytes := s.getBaseDeviceSize(c)
+
+	newBasesizeBytes := 53687091200 //50GB in bytes
+
+	if int64(newBasesizeBytes) < oldBasesizeBytes {
+		c.Skip(fmt.Sprintf("New base device size (%v) must be greater than (%s)", units.HumanSize(float64(newBasesizeBytes)), units.HumanSize(float64(oldBasesizeBytes))))
+	}
+
+	if err := s.d.Restart("--storage-opt", fmt.Sprintf("dm.basesize=%d", newBasesizeBytes)); err != nil {
+		c.Fatalf("we should have been able to start the daemon with increased base device size: %v", err)
+	}
+
+	basesizeAfterRestart := s.getBaseDeviceSize(c)
+	newBasesize, err := convertBasesize(newBasesizeBytes)
+	if err != nil {
+		c.Fatalf("Error in converting base device size: %v", err)
+	}
+
+	if newBasesize != basesizeAfterRestart {
+		c.Fatalf("Basesize passed is not equal to Basesize set")
+	}
+}
+
+func convertBasesize(BasesizeBytes int) (int64, error) {
+	Basesize := units.HumanSize(float64(BasesizeBytes))
+	Basesize = strings.Trim(Basesize, " ")[:len(Basesize)-3]
+	BasesizeFloat, err := strconv.ParseFloat(strings.Trim(Basesize, " "), 64)
+	if err != nil {
+		return 0, err
+	}
+	return int64(BasesizeFloat) * 1024 * 1024 * 1024, nil
+}
+
+func (s *DockerDaemonSuite) getBaseDeviceSize(c *check.C) int64 {
```

Comment:
we probably want to move this to `docker_utils.go` and use `/info` endpoint instead of grep?